sklearn/metrics/pairwise.py

```diff
@@ -1556,9 +1556,9 @@ def distance_metrics():
     return PAIRWISE_DISTANCE_FUNCTIONS
 
 
-def _dist_wrapper(dist_func, dist_matrix, slice_, *args, **kwargs):
-    """Write in-place to a slice of a distance matrix."""
-    dist_matrix[:, slice_] = dist_func(*args, **kwargs)
+def _dist_wrapper(dist_func, slice_, *args, **kwargs):
+    """Computing a slice of distance matrix."""
+    return dist_func(*args, **kwargs)
```

Comment:
This changes the logic for the default threading backend from writing in-place to returning the ndarray to the caller.

May you run a quick benchmark with `%%timeit` comparing how long it takes to run `pairwise_distances` on `main` vs this PR with default threading backend and a moderately sized dataset?

Reply:
Thanks Thomas,
it is indeed faster with the write-in-place approach. This is the test I run:

```
import numpy as np
from sklearn.metrics import pairwise, pairwiseWriteInPlace
import timeit

x = np.arange(0, 70, .5)
y = np.arange(500, 650, .5)
npt = len(x)*len(y)

x, y = np.meshgrid(x, y)

pts = np.zeros([npt, 2])
pts[:,0] = x.flatten()
pts[:,1] = y.flatten()

def pairwise_distances():
    pairwise.pairwise_distances(pts, metric='euclidean', n_jobs=2)

def pairwise_distancesWriteInPlace():
    pairwiseWriteInPlace.pairwise_distances(pts, metric='euclidean', n_jobs=2)

t1 = timeit.timeit(pairwise_distances, number=2)
print(f"return ndarray exec time: {t1}")

t2 = timeit.timeit(pairwise_distancesWriteInPlace, number=2)
print(f"write in place exec time: {t2}")
```

The result is
return ndarray exec time: 57.272275117982645
write in place exec time: 40.82838660501875

Maybe we could use 2 different dealyed on the basis of the beckend? Or do you have some better strategy to suggest?

Reply:
I amended the PR to using a write-in-place approach in a multithreading backend, and a return-to-caller approach in multiprocessing. Now the timeit test gives the same scores for the old and the new codes